### PR TITLE
server: improve the memory usage of schemaHandler's ServeHTTP to avoid OOM

### LINKED ISSUE #30969: improve the memory usage of schemaHandler's ServeHTTP
## Enhancement

Deploy a small cluster, and create 10000 tables, each table has 400 columns.
The memory usage of schemaHandler's ServeHTTP is very hight and may cause OOM.
Here is the grafana:
image

And you can see that TiDB would OOM continually.
image

### PR DESCRIPTION
**Pull request by wjhuang2016**
### What problem does this PR solve?

Issue Number: close #30969

Problem Summary:
Previously, we call writeData() for all the databases's tables. And it would use too much memory and can't deallocate it quickly.
If the database has too many tables, it may cause OOM.

### What is changed and how it works?
call writeData() for every table. 

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
Replace this PR's TiDB at 16:30, and you can see that the memory is deallocated quickly. TiDB server is not OOM any more.
That one restart later is caused by replacing TiDB again and it's expected.
image
image

- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
improve the memory usage of schemaHandler's ServeHTTP to avoid OOM
```

### CODE REVIEW

**Review by djshow832 — COMMENTED** [review]
Is it covered by any test cases? How do you guarantee that the result format is the same as before?

File: server/http_handler.go [hunk 1]
```diff
@@ -1083,11 +1083,9 @@ func (h schemaHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 		// all table schemas in a specified database
 		if schema.SchemaExists(cDBName) {
 			tbs := schema.SchemaTables(cDBName)
-			tbsInfo := make([]*model.TableInfo, len(tbs))
-			for i := range tbsInfo {
-				tbsInfo[i] = tbs[i].Meta()
+			for _, tb := range tbs {
+				writeData(w, tb.Meta())
```
Comment (lance6716) [inline server/http_handler.go:1087]: should we write `[]*model.TableInfo{tb.Meta()}` to keep compatibility ?

**Comment by wjhuang2016** [comment]
> Is it covered by any test cases? How do you guarantee that the result format is the same as before?

NO, I think it should be covered by the system test?

File: server/http_handler.go (hunk 1 above)
Comment (wjhuang2016) [reply server/http_handler.go:1087]: Would it break compatibility? I tried KeyVisualizer and it worked.

Comment (wjhuang2016) [reply server/http_handler.go:1087]: Good catch, I will fix it

**Comment by wjhuang2016** [comment]
> Is it covered by any test cases? How do you guarantee that the result format is the same as before?

It's covered by TestGetSchema, I used @lance6716's method and it worked.

**Comment by tangenta** [comment]
How about asserting the number of `TableInfo` > 2?

https://github.com/pingcap/tidb/blob/05b99603901ae0981aa10027dc95e4d595a54c1e/server/http_handler_test.go#L871-L876

**Review by djshow832 — COMMENTED** [review]
What about `writeData(w, schema.AllSchemas())`?

**Comment by zhangjinpeng1987** [comment]
How about extracting the logic to an individual function and adding a unit test?

**Comment by wjhuang2016** [comment]
> What about `writeData(w, schema.AllSchemas())`?

@djshow832 I think it's fine for now. Since a database's json is small. It may become problem if there are more than 10 000 databases?

**Comment by wjhuang2016** [comment]
> > What about `writeData(w, schema.AllSchemas())`?
> 
> @djshow832 I think it's fine for now. Since a database's json is small. It may become problem if there are more than 10 000 databases?

I tested it with 10 000 databases, not obvious memory increment.

**Review by djshow832 — COMMENTED** [review]
When there are 10000 tables, Key Viz fetches data with timeout and retries.
Does this PR also speed it up, so Key Viz has no timeout anymore?

**Comment by wjhuang2016** [comment]
> When there are 10000 tables, Key Viz fetches data with timeout and retries. Does this PR also speed it up, so Key Viz has no timeout anymore?

Yes,
image

**Comment by wjhuang2016** [comment]
> When there are 10000 tables, Key Viz fetches data with timeout and retries. Does this PR also speed it up, so Key Viz has no timeout anymore?

@djshow832 I checked the log and found that it's timeout. So we need to fix the timeout on the dashboard side.

**Review by djshow832 — APPROVED** [review]

File: server/http_handler_test.go [hunk 2]
```diff
@@ -873,7 +876,7 @@ func TestGetSchema(t *testing.T) {
 	err = decoder.Decode(&lt)
 	require.NoError(t, err)
 	require.NoError(t, resp.Body.Close())
-	require.Greater(t, len(lt), 0)
+	require.Equal(t, len(lt), 3)
```
Comment (tangenta) [inline server/http_handler_test.go:879]: Maybe greater than 2 is enough? Otherwise it is easy to break if someone add tables to `tidb` in test.

**Review by tangenta — APPROVED** [review]

**Pull request merged**